### PR TITLE
Panzer tangent unit tests (Scatter Tpetra)

### PR DESCRIPTION
**Pull request by reuterb**
@trilinos/panzer 

## Motivation

This is the second MR related to Panzer's tangent capability (see #13576).
Here we refresh the `ScatterResidual_Tpetra` and add tangent capabilities on device.
We also create a unit test, noting that a previously existing test of the same name actually tested the blocked version of the evaluator.
That test has been renamed appropriately and that evaluator `ScatterResidual_BlockedTpetra` will be the focus of the next MR.

## Related Issues

* Closes `put-issue-number-here`

### CODE REVIEW

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterResidual_Tpetra_impl.hpp [hunk 1]
```diff
@@ -191,50 +203,30 @@ preEvaluate(typename TRAITS::PreEvalData d)
...
-            LO lid = LIDs[offset];
-            ScalarT value = (scatterFields_[fieldIndex])(worksetCellIndex,basis);
-            for(int d=0;d<value.size();d++)
-              dfdp_vectors_[d][lid] += value.fastAccessDx(d);
-         }
-      }
-   }
+  dfdpFieldsVoV_.syncHostToDevice();
+
+  // TODO BWR DO WE WANT TO BE ABLE TO FILL RESIDUAL TOO?
```
Comment (reuterb) [inline packages/panzer/disc-fe/src/evaluators/Panzer_ScatterResidual_Tpetra_impl.hpp:221]: @rppawlo Please confirm that we do want to be able to do this. It wasn't that way before but the Jacobian version does it.

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterResidual_Tpetra_impl.hpp [hunk 2]
```diff
@@ -191,50 +203,30 @@ preEvaluate(typename TRAITS::PreEvalData d)
   std::vector<std::string> activeParameters =
     rcp_dynamic_cast<ParameterList_GlobalEvaluationData>(d.gedc->getDataObject("PARAMETER_NAMES"))->getActiveParameters();
 
-  dfdp_vectors_.clear();
+  dfdpFieldsVoV_.initialize("ScatterResidual_Tpetra<Tangent>::dfdpFieldsVoV_",activeParameters.size());
+
   for(std::size_t i=0;i<activeParameters.size();i++) {
+    // TODO BWR DO NOT UNDERSTAND THIS... Why is param->f dfdp?
```
Comment (reuterb) [inline packages/panzer/disc-fe/src/evaluators/Panzer_ScatterResidual_Tpetra_impl.hpp:209]: Another

File: packages/panzer/disc-fe/src/evaluators/Panzer_ScatterResidual_Tpetra_impl.hpp [hunk 3]
```diff
@@ -483,6 +512,36 @@ evaluateFields(typename TRAITS::EvalData workset)
...
+  functor.r_data = r->getLocalViewDevice(Tpetra::Access::ReadWrite);
+  functor.lids = scratch_lids_;
+
+  // for each field, do a parallel for loop
+  for(std::size_t fieldIndex = 0; fieldIndex < scatterFields_.size(); fieldIndex++) {
+    functor.offsets = scratch_offsets_[fieldIndex];
+    functor.field = scatterFields_[fieldIndex];
+    functor.dfdp_fields = dfdpFieldsVoV_.getViewDevice();
+    // TODO BWR is there a better way? Is this correct?
+    functor.num_params = PHX::getFadSize(scatterFields_[fieldIndex].get_static_view())-1;
```
Comment (reuterb) [inline packages/panzer/disc-fe/src/evaluators/Panzer_ScatterResidual_Tpetra_impl.hpp:539]: @rppawlo Is this a good way to get the number of derivatives?

Comment (rppawlo) [reply packages/panzer/disc-fe/src/evaluators/Panzer_ScatterResidual_Tpetra_impl.hpp:539]: @reuterb This is usually what I use
```
const int fadSize = Kokkos::dimension_scalar(field_.get_view());
```

**Comment by reuterb** [comment]
@rppawlo Should be ready for review when you get a chance

**Review by rppawlo — APPROVED** [review]

**Pull request merged**